### DISCUSSION TITLE
String Gauge

### DISCUSSION THREAD

**Discussion by BFergerson** [apache/skywalking#9648] [Q&A] [answered]
How would you implement something like a string gauge? I'd like to keep track of the latest instance of a string value. Something like a "newest member" gauge. For example, many forums have a "Welcome newest member: $username", like so:
![image](https://user-images.githubusercontent.com/3278877/191331961-648d3aae-c10c-4485-a5af-489ed3676594.png)

I know I can use traces, logs, and perhaps events, but those feel like overkill. I don't need every string. One string per `meter.report_interval` would suffice. What's an ideal way to do that?

**Comment by wu-sheng** [comment, marked as answer]
I think we don't have proper way to support this. 
Is this working for APM scenario? Newest member seems a business monitoring?

**Reply by BFergerson** [reply]
Newest member was probably a bad example. There is a quote [here](https://github.com/elastic/apm/issues/287#issuecomment-651652972) in a discussion on the Elastic APM that provides a few more example uses. I want to track string ids and usernames felt close to my use case.

Thinking about it a bit more, I don't think gauge would be the appropriate tool to do it though. If a gauge is a metric that can increase and decrease then adding a string there wouldn't make sense. What would be your thoughts on something like a [StringMonitor](https://docs.oracle.com/javase/9/docs/api/javax/management/monitor/StringMonitor.html)? Or simply SerializableMonitor?

**Reply by wu-sheng** [reply]
This seems more like a status monitoring? The string shows the status description rather than a number code?
I am wondering how to visual this? A status flow graph, or statistics of status in a group?(such as pods of a service?

**Reply by wu-sheng** [reply]
StringMonitoring makes sense at source codes level. Do you want to do some debugging monitoring? Keep tracking a field value, or something similar?

**Reply by BFergerson** [reply]
Looks like DropWizard Metrics supports something akin to [SerializableGauge](https://www.javadoc.io/doc/io.dropwizard.metrics/metrics-core/4.2.0/com/codahale/metrics/Gauge.html).

**Reply by BFergerson** [reply]
Didn't see your messages when I commented. Yeah, that's exactly what I'm trying to do. I want to monitor variable values with Source++ and I'm trying to figure out how to do that efficiently. So I know it's more of a my issue than a your issue but I'm curious how I can use SkyWalking to do it. The best idea I have right now is to store it in multiple double values and then decode the doubles back into a string, which won't be very pretty.

**Reply by wu-sheng** [reply]
I wouldn't be surprised there are some projects have. I just want to discuss what we could do typically in our agents to provide typical use cases.
Currently, all adopted concepts we added have been used in default monitoring and dashboard.

**Reply by BFergerson** [reply]
Fair enough. If no one else needs this functionality I won't impose it on SkyWalking. I think for my use case I'll use logs or events, thanks.

**Reply by wu-sheng** [reply]
If you have ideas of how to use this in general monitoring, this should be easy to add from tech perspective.

**Comment by Superskyyy** [comment]
I feel like there are some unique scenarios that could be also useful apart from source++ (something like an event but as a stream), maybe IOT? I'll update here if I think of something concrete.
